### ISSUE TITLE
macOS Text Selection: shift+left should expand selection

### ISSUE THREAD

**Issue by troughton** [microsoft/vscode#79104]
- VSCode Version: 1.37.0
- OS Version: macOS Mojave 10.14.6

On macOS, the standard text selection behaviour is to expand the current selection when shift + the left arrow key is pressed:

> Here is the **example** text with 'example' selected (e.g. by double-clicking the text)
> Here is th**e example** text on macOS in most applications after pressing shift + the left arrow key twice.
> Here is the **exampl**e text on macOS in Visual Studio code after the same action.

To be consistent with platform conventions, shift + the left arrow should ideally expand the selection. I believe the current behaviour matches the behaviour on Windows.

If there is already a configuration option that changes this behaviour, then I would suggest that the default on macOS be changed to the standard system behaviour.

**Comment by alexandrudima** [comment]
@troughton I'm not sure I understand your steps. For me, shift + left arrow expands the selection:

![Kapture 2019-08-14 at 17 57 24](https://user-images.githubusercontent.com/5047891/63036261-fed4d180-bebc-11e9-9c19-9104945364a2.gif)

Can you please check your `keybindings.json`, maybe you have added a keybinding for `shift+left`? Does this reproduce when disabling all extensions?

**Comment by troughton** [comment]
To be clear, starting with no selection behaves correctly for me. However, if you start by double-clicking a word to select it, shift + left will deselect the last letter in the word rather than expanding the selection one letter to the left.

**Comment by alexandrudima** [comment]
@troughton I cannot reproduce.

Can you please check your `keybindings.json`, maybe you have added a keybinding for `shift+left`? Does this reproduce when disabling all extensions?

**Comment by troughton** [comment]
Yes, it reproduces with all extensions disabled, and no, I have nothing custom in `keybindings.json`. Here are some clips to more clearly show what I mean:

![ShiftSelection-VSCode](https://user-images.githubusercontent.com/2133470/63088695-c828b500-bfa9-11e9-9389-52947750a306.gif)

![ShiftSelection-TextEdit](https://user-images.githubusercontent.com/2133470/63088707-cfe85980-bfa9-11e9-9345-6021e8fad4b4.gif)

**Comment by alexandrudima** [comment]
Oh, interesting. I think what is different is that we have the concept of a cursor even when there is a selection. You can see us rendering it and the selection always grows/shrinks relative to the cursor. This is consistent with other code editors.

I will leave this open as a feature request, but I don't think I've seen any code editor have the behavior of TextEdit...

**Comment by troughton** [comment]
I don’t think it’s the cursor that’s different - in your earlier example where you started with no initial selection the selection expanded to the left. I’m just guessing here, but it seems like the selection starts from an anchor point and always expands from/contracts towards it. In the current behaviour, that anchor point is chosen to be the start of the word, meaning shift-left deselects towards it. On macOS/AppKit, however, there’s an anchor _range_ instead, and you can never deselect anything within the initial anchor range.

In terms of other code editors - pretty much anything using standard system controls on macOS behaves like TextEdit, including Xcode. That’s the main reason I filed this bug – when switching between Xcode and VS Code I kept having my muscle memory broken.

I understand that this is going to be a low priority bug, though.

**Comment by alexandrudima** [comment]
Just to double check, I have also tried on Windows, and on Windows the behavior is the one currently implemented in VS Code.

Chrome respects the macOS behavior, so it behaves differently on macOS than on Windows.

**Comment by stormsilver** [comment]
I keep getting bit by this when I double-click to select a variable, then shift+left to select the $ in front of the variable. I kept wondering why I would end up pasting the variable with neither the preceding $ or the last character of the variable name. It took me a long time to figure out what was happening because this is ingrained so deeply in muscle memory.

It's worth noting that shift+right behaves as expected by expanding the selection to the right. It's only shift-left that has the wrong behavior.

> anchor point is chosen to be the opposite end of wherever you first try to shift-expand after the initial selection
This is the proper behavior.

**Comment by beelsebob** [comment]
My issue was duped to here.  Just adding to the pile on.  Since there was a comment above about "not having seen a code editor that behaves like this", here's a list of code editors on macOS that behave like this:

• Xcode
• TextMate
• CodeEdit
• BBEdit
• Atom
• Phoenix Code
• Noto
• SubEthaEdit
• CotEdit

In checking these, the only ones I've found that didn't behave like the macOS default are SublimeText and VSCode (and ofc weird outliers like vim); so this isn't an obscure behaviour even in code editors.
